File: packages/mobile-ios/Sources/RTCExpress/TokenClient.swift
import Foundation

public enum TokenClient {
    public static func fetchToken(serverUrl: String, request: TokenRequest) async throws -> TokenResponse {
        let url = URL(string: "\(serverUrl.trimmingCharacters(in: .init(charactersIn: "/")))/v1/token")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try JSONEncoder().encode(request)

        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let http = response as? HTTPURLResponse, http.statusCode < 300 else {
            let err = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw NSError(domain: "RTCExpress", code: 1, userInfo: [NSLocalizedDescriptionKey: err ?? "Token request failed"])
        }
        return try JSONDecoder().decode(TokenResponse.self, from: data)
    }
}

File: packages/mobile-android/sdk/src/main/java/com/rtcexpress/sdk/TokenClient.kt
package com.rtcexpress.sdk

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

class TokenClient(private val http: OkHttpClient = OkHttpClient()) {
    fun fetchToken(serverUrl: String, request: TokenRequest): TokenResponse {
        val body = JSONObject()
            .put("appId", request.appId)
            .put("appSecret", request.appSecret)
            .put("userId", request.userId)
        request.roomId?.let { body.put("roomId", it) }

        val httpRequest = Request.Builder()
            .url("${serverUrl.trimEnd('/')}/v1/token")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        http.newCall(httpRequest).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                val err = runCatching { JSONObject(text).optString("error") }.getOrNull()
                throw IllegalStateException(err ?: "Token request failed (${response.code})")
            }
            val json = JSONObject(text)
            return TokenResponse(
                token = json.getString("token"),
                expiresIn = json.optInt("expiresIn", 3600)
            )
        }
    }

    fun fetchPlatformConfig(serverUrl: String, appId: String? = null): JSONObject {
        val url = buildString {
            append(serverUrl.trimEnd('/'))
            append("/v1/config")
            if (!appId.isNullOrBlank()) append("?appId=").append(appId)
        }
        val request = Request.Builder().url(url).get().build()
        http.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (!response.isSuccessful) throw IllegalStateException("Config request failed")
            return JSONObject(text)
        }
    }
}
